content/oci/oci.go

```diff
@@ -398,6 +402,49 @@ func (s *Store) writeIndexFile() error {
 	return os.WriteFile(s.indexPath, indexJSON, 0666)
 }
 
+// GC removes garbage from Store. The garbage to be cleaned are:
+//   - unreferenced (dangling) blobs in Store which have no predecessors
+//   - garbage blobs in the storage whose metadata is not stored in Store
+func (s *Store) GC(ctx context.Context) error {
+	s.sync.Lock()
+	defer s.sync.Unlock()
+
+	// clean up dangling layers in Store
+	danglings := s.graph.GetDanglingLayers()
+	for _, desc := range danglings {
+		// do not remove existing manifests in the index
+		if _, err := s.tagResolver.Resolve(ctx, string(desc.Digest)); err == errdef.ErrNotFound {
+			// remove the blob and its metadata from the Store
+			if err := s.deleteNode(ctx, desc); err != nil {
+				return err
+			}
+		}
+	}
+
+	// clean up garbage blobs in the storage
+	rootpath := filepath.Join(s.root, "blobs")
+	return filepath.Walk(rootpath, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		// skip the directories
+		if !info.IsDir() {
+			alg := filepath.Base(filepath.Dir(path))
+			blobDigest, err := digest.Parse(fmt.Sprintf("%s:%s", alg, info.Name()))
+			if err != nil {
+				return err
+			}
+			if exists := s.graph.Exists(blobDigest); !exists {
+				// remove the blob from storage if it does not exist in Store
+				if err := s.storage.deleteByDigest(ctx, blobDigest); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
```

Comment:
nit: code style: consider
```go
if info.IsDir() {
    return nil
}

// do other stuffs
return nil
```

Reply:
We should scan every `alg` directory and every `encoded` in it, rather than walking into every nested sub-dir under "blobs" directory. For example, given the below structure:

```
blobs
  |- sha256
      |- <encoded_1>
      |- random_string
 |- sha512
     |- <encoded_2>
     |- random_dir_1
 |- random_dir_2
```

We should not walk into `random_dir_1` and `random_dir_2`.

Reply:
> nit: code style: consider
> 
> ```go
> if info.IsDir() {
>     return nil
> }
> 
> // do other stuffs
> return nil
> ```

changed accordingly.

Reply:
> We should scan every `alg` directory and every `encoded` in it, rather than walking into every nested sub-dir under "blobs" directory. For example, given the below structure:
> 
> ```
> blobs/
>   |- sha256/
>      |- <encoded_1>
>      |- random_string
>   |- sha512/
>      |- <encoded_2>
>      |- random_dir_1/
>   |- random_dir_2/
> ```
> 
> We should not walk into `random_dir_1/` and `random_dir_2/`.

I think this can be achieved by the below? If the parse fails the invalid content will be silently skipped.

```
			blobDigest, err := digest.Parse(fmt.Sprintf("%s:%s", alg, info.Name()))
			if err != nil {
				return err
			}
```

Reply:
Replaced `Walk` with a loop. This comment is no longer relevant.